### ISSUE TITLE
raft: `panic: group 1 200000002 state.commit 9894 is out of range [10, 9893]`

### ISSUE THREAD

**Issue by mrtracy** [cockroachdb/cockroach#2592]
This occurred on a six-node cluster which was under constant insertion load; splits were occuring and ranges were being rebalanced.  At some point the node which was receiving the insertion load crashed with this panic, and it re-occured immediately on every attempt to restart the node.

I think this is different than #1878, but I'm not sure.

Unfortunately, I do not have a test which currently reproduces this; working on that.  Detailed stack:

```
E0917 23:39:49.355218 3663 raft/raft.go:777 group 1 200000002 state.commit 9894 is out of range [10, 9893]
panic: group 1 200000002 state.commit 9894 is out of range [10, 9893]

goroutine 44 [running]:
github.com/cockroachdb/cockroach/multiraft.(*raftLogger).Panicf(0xc8202e7d80, 0x4e92580, 0x2b, 0xc8202b88c0, 0x4, 0x4)
/Users/matt/go/src/github.com/cockroachdb/cockroach/multiraft/raft.go:121 +0x1a2
github.com/coreos/etcd/raft.(*raft).loadState(0xc8202ec270, 0xf, 0x300000003, 0x26a6, 0x0, 0x0, 0x0)
/Users/matt/go/src/github.com/coreos/etcd/raft/raft.go:777 +0x2b3
github.com/coreos/etcd/raft.newRaft(0xc820302540, 0xc82016b440)
/Users/matt/go/src/github.com/coreos/etcd/raft/raft.go:206 +0x8f0
github.com/coreos/etcd/raft.(*multiNode).run(0xc82016b0e0)
/Users/matt/go/src/github.com/coreos/etcd/raft/multinode.go:183 +0x379
created by github.com/coreos/etcd/raft.StartMultiNode
/Users/matt/go/src/github.com/coreos/etcd/raft/multinode.go:56 +0x30d
```

**Comment by bdarnell** [comment]
This is probably #768 and will be fixed by replica tombstones.

**Comment by es-chow** [comment]
The same panic had reproduced in our testing (single node, single store, single range, single replica). It happened in the following scenario:

1. Another panic happened just after `multiraft.writeTask` finished `group.SetHardState` call but before `group.Append` call. So the `state.commit` had been stabled to the storage but the Entries had not been stabled to the storage.

2. That roachNode got restarted, when `raft.newRaft` try to `loadState`, `state.commit` will be 1 greater than `lastIndex`.

The cause for this issue may be the `group.SetHardState` and `group.Append` is not atomic. But I am not sure if the mentioned six node issue come from the same cause.

In etcd implementation `etcd/etcdserver/raft.go`, it call `r.storage.Save(rd.HardState, rd.Entries)` to store the HardState and Entries atomically. It's not the same in multiraft. Should we follow the same pattern as etcd?

Another doubt is if ApplySnapshot and SetHardState is not atomic, anything bad will happen? If that can happen, we should make all three in atomic.

**Comment by bdarnell** [comment]
@es-chow, what was the first panic between `group.SetHardState` and `group.Append`? Was it the same "commit out of range" or something else? 

I think this panic in `etcd/raft` is a little too aggressive, and it might be good to relax it (so that a node can be aware that the cluster has committed entries that it doesn't know about yet. We'd basically move checks like [this](https://github.com/coreos/etcd/blob/c645ac23c0093e2b0a93fa5f07a947344d7ef779/raft/raft.go#L308) to the point where `HardState.Commit` is used, instead of when the commit index is broadcast). 

I think a quick fix for this issue would be to reorder the write operations: snapshot, entries, and HardState in that order. They don't have to be atomic, they just have to all be finished before any messages are sent (and apparently HardState should not be written unless the other two are). Putting these in an atomic batch would be good for performance but is not necessary for correctness. Instead of following etcd's example with a method that takes `(HardState, []Entry)`, we should have explicit batching support in the `Storage`/`WriteableGroupStorage` interfaces so we can batch things across all groups, not just within one.

**Comment by es-chow** [comment]
@bdarnell, The first panic is not "commit out of range", it's a "Event buffer full" in `feed.Publish` when we test use an older version.

Following your "We'd basically move checks like this to", "this" is not pointer to any checks?

Thanks for your suggestion, we can implementation that batching.

**Comment by bdarnell** [comment]
The check I'm referring to is the use of `min()`. The leader guarantees that `Message.Commit` only points to an entry that the follower has acknowledged (or is being appended by the same message, in the `MsgApp` case).

**Comment by BramGruneir** [comment]
Closed via #3044
